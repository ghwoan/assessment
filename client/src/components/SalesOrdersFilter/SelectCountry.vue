<template>
<select v-model="selected" @change="onChange">
   <option value="-">All</option>
   <option v-for="country in countryList" :key="country.country" :value="country.country">
   {{country.country}}</option>
</select>
</template>

<script>
import { getCountryList } from '../../data/salesorders.js';

export default {
   props: {
    value: String
   },
   data() {
      return {
         selected: this.value,
         countryList: []
      };
   },
   created() {
      this.fetchCountryList();
   },
   mounted(){

   },
   watch: {
      value: function(newVal){
         console.log(newVal);
         this.selected = newVal;
      }
   },
   methods:{
      //fetch country list from the server
      fetchCountryList: function() {
         getCountryList()
            .then(response => {

               if (response && response.data) {
                  this.countryList= response.data.data;
               }
            })
            .catch(e => {
               console.log(e); 
         })
      },
      //handle on country selection change
      onChange: function(event){
         this.$emit("valueChange", event.target.value);
      }
   }
}
</script>
<style scoped>

</style>
