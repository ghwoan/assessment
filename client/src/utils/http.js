import axios from 'axios';

const defConfig = {
  baseURL: process.env.VUE_APP_SERVER_URL || `http://localhost:8080/`,
  headers: {
    Authorization: process.env.VUE_APP_API_KEY || 'Testing'
  }
};

export const http = axios.create(defConfig);