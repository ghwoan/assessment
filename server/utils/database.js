import mysql from 'mysql2';
import dbConfig from "../config/db.config.js";

const pool = mysql.createPool({
    host: dbConfig.host,
    user: dbConfig.USER,
    database: dbConfig.DB,
   password: dbConfig.PASSWORD,
   waitForConnections: true,
   connectionLimit: 20,
});

const db = pool.promise();

export default db;