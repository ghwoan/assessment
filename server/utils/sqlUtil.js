
export const createDateRangeFilter = (fieldName, prefix, startDate, endDate) => {
   let sql = "";
   if (startDate) {
      sql = fieldName + " >= '" + startDate + "'";
   }
   if (endDate) {
      if (sql) {
         sql = sql + " AND ";
      }
      sql = sql + fieldName + " <='" + endDate + "'";
   }
   if (sql) {
      sql = " (" + sql + ")";
      if (prefix) {
         sql = prefix + " AND " + sql;
      }
   } else {
      sql = prefix;
   }
   return sql;
}

//define enum for FieldType
export const FieldTypeEnum = {
   NUMBER : 1,
   STRING : 2
}

export const escapeChar = (value) => {
   if (!value) {
      return value;
   } else {
      return "'" + value.split("'").join("\\'") + "'";
   }
}

export const createOpFilter = (fieldName, prefix, opFilter, fieldType) => {
   let sql = prefix;
   let items = "";
  
   if (!opFilter || !opFilter.list) {
      return prefix;
   };
   console.log(opFilter.list);
   if (opFilter.list.length) {
      opFilter.list.forEach(c => {
         if (items) items = items + ",";
         if (fieldType == FieldTypeEnum.NUMBER) {
            items = items + c;   
         } else {
            items = items +  escapeChar(c) ;
         }
      });
   }
   if (items) {
      items = "(" + items + ")";
      if (opFilter.op == "IN") {
         items = fieldName + " IN " + items;
      } else[
         items = fieldName + " NOT IN " + items
      ]
      sql = sql  + " AND (" + items + ")";
   }
   return sql;
}