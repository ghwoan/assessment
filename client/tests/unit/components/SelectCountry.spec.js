import { http } from '@/utils/http';
import { shallowMount, mount } from '@vue/test-utils';
import SelectCountry from '../../../src/components/controls/SelectCountry.vue';


describe('SelectCountry.vue', () => {
  xit('renders SelectCountry empty countryList', () => {
    const msg = 'All';
    const wrapper = shallowMount(SelectCountry, {
      props: {  value: "China"}
    })
    console.log(wrapper.html());
    expect(wrapper.props().value).toBe('China');
    expect(wrapper.text()).toMatch(msg);
  })

  xit('renders SelectCountry with countryList', async () => {
    const wrapper = await mount(SelectCountry, {
      props: {  value: "China"}
    });
    await wrapper.setData({ selected: "Indonesia", countryList: [{ country: "Indonesia" }, { country: "China" }]});
    
   // expect(wrapper.text()).toMatch(msg);
    expect(wrapper.html()).toMatchSnapshot();
   
  })

  //a mock return of the getCountryList
  http.get = jest.fn().mockResolvedValue(
    { data: { data: [{ country: "Malaysia" }, { country: "Singapore" }] } });
  
  xit('renders SelectCountry with countryList2', async () => {
    const wrapper = await mount(SelectCountry, {
      props: {  value: "Malaysia"}
    });
    console.log(wrapper.html());
    expect(wrapper.html()).toMatchSnapshot();
   
  })
})
