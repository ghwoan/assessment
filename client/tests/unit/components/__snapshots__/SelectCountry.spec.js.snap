// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectCountry.vue renders SelectCountry with countryList 1`] = `
<select>
  <option value="-">All</option>
  <option value="Indonesia">Indonesia</option>
  <option value="China">China</option>
</select>
`;

exports[`SelectCountry.vue renders SelectCountry with countryList2 1`] = `
<select>
  <option value="-">All</option>
  <option value="Malaysia">Malaysia</option>
  <option value="Singapore">Singapore</option>
</select>
`;
